ksm.go

```diff
@@ -0,0 +1,214 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+package main
+
+import (
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"syscall"
+)
+
+type ksmSetting struct {
+	pagesPerScanFactor uint32
+	scanIntervalMs     uint32
+}
+
+const (
+	ksmSlug       = "slug"
+	ksmSlow       = "slow"
+	ksmStandard   = "standard"
+	ksmAggressive = "aggressive"
+)
+
+var ksmSettings = map[string]ksmSetting{
+	ksmSlug:       {1000, 500}, // Every 500ms, we scan 1 page for every 1000 pages available in the system
+	ksmSlow:       {500, 100},  // Every 100ms, we scan 1 page for every 500 pages available in the system
+	ksmStandard:   {100, 10},   // Every 10ms, we scan 1 page for every 100 pages available in the system
+	ksmAggressive: {10, 1},     // Every ms, we scan 1 page for every 10 pages available in the system
+}
+
+const (
+	ksmRoot           = "/sys/kernel/mm/ksm/"
+	ksmRunFile        = "run"
+	ksmPagesToScan    = "pages_to_scan"
+	ksmSleepMillisec  = "sleep_millisecs"
+	ksmRun            = "1"
+	defaultKSMSetting = ksmAggressive
+)
+
+type sysfsAttribute struct {
+	path string
+	file *os.File
+}
+
+func (attr *sysfsAttribute) open() error {
+	file, err := os.OpenFile(attr.path, os.O_RDWR|syscall.O_NONBLOCK, 0660)
+	attr.file = file
+	return err
+}
+
+func (attr *sysfsAttribute) close() error {
+	err := attr.file.Close()
+	attr.file = nil
+
+	return err
+}
+
+func (attr *sysfsAttribute) read() (string, error) {
+	_, err := attr.file.Seek(0, os.SEEK_SET)
+	if err != nil {
+		return "", err
+	}
+
+	data, err := ioutil.ReadAll(attr.file)
+	if err != nil {
+		return "", err
+	}
+
+	return string(data), nil
+}
+
+func (attr *sysfsAttribute) write(value string) error {
+	_, err := attr.file.Seek(0, os.SEEK_SET)
+	if err != nil {
+		return err
+	}
+	_, err = attr.file.WriteString(value)
+
+	return err
+}
+
+// store a sysfs attribute and returns its old value
+func (attr *sysfsAttribute) store(value string) (string, error) {
+	if attr.file == nil {
+		err := attr.open()
+		if err != nil {
+			return "", err
+		}
+
+		defer func() {
+			_ = attr.close()
+		}()
+	}
+
+	oldValue, err := attr.read()
+	if err != nil {
+		return "", err
+	}
+
+	err = attr.write(value)
+	if err != nil {
+		return "", err
+	}
+
+	return oldValue, nil
+}
+
+func isKSMAvailable(ksmDir string) error {
+	info, err := os.Stat(ksmDir)
+	if err != nil || !info.IsDir() {
+		return fmt.Errorf("%s is not available", ksmDir)
+	}
+
+	return nil
+}
+
+func pagesToScan(s ksmSetting) (string, error) {
```

Comment:
Shame we are not using the 's' field here - I know in a later patch we add a more complex bit of math to scale against system pages size. Had me scratching my head a little thinking 'but we have a table, are we not using it?' - but, I see we do use the table for the delay timing.
Just a nitpick :-)